misk/src/main/kotlin/misk/web/actions/UpstreamResourceInterceptor.kt

```diff
@@ -0,0 +1,49 @@
+package misk.web.actions
+
+import misk.web.NetworkChain
+import misk.web.NetworkInterceptor
+import misk.web.Response
+import misk.web.forwardRequestTo
+import okhttp3.HttpUrl
+
+class UpstreamResourceInterceptor(
+  private val mappings: MutableList<out Mapping>
+) : NetworkInterceptor {
+  @Suppress("UNUSED_PARAMETER")
+  override fun intercept(chain: NetworkChain): Response<*> {
+    var matchedMapping = findMappingMatch(chain) ?: return chain.proceed(chain.request)
+    val proxyUrl = HttpUrl.parse(
+        matchedMapping.upstreamBaseUrl.toString() + chain.request.url.encodedPath().removePrefix(
+            matchedMapping.localPathPrefix))!!
+    return chain.request.forwardRequestTo(proxyUrl)
+  }
+
+  private fun findMappingMatch(chain: NetworkChain): Mapping? {
+    var matchedMapping: Mapping? = null
+    for (mapping in mappings) {
+      if (!chain.request.url.encodedPath().startsWith(mapping.localPathPrefix)) continue
+      if (matchedMapping == null || mapping.localPathPrefix.count { ch -> ch == '/' } > matchedMapping.localPathPrefix.count { ch -> ch == '/' }) matchedMapping =
```

Comment:
it looks like you’re trying to handle the case where multiple mappings match. I think we should forbid this and just return the first path that matches

Maybe we should confirm that none of the mappings’ localPathPrefix is a prefix of any others

Reply:
I'd be okay forbidding outright but maximal matching is likely necessary for the following case.

`/_admin/*`  -> `localhost:3001`: generic mapping for all _admin actions
`/_admin/config/` -> `localhost:3001`: the index page for the `config` module
`/_admin/config/all` -> `localhost:8080`: all `config` domain API calls should be handled by Misk, not forwarded to webpack

Options
1) Leave rules as maximal matching, though would need to be well documented and potentially add in `*` matching so all API calls don't need to be explicitly specified as a matching.
2) Enforce non-overlapping first prefix matching and re-attempt all failed forwarding URLs on Misk server. Potential problem is how to handle redirecting `/_admin/`which contains the main `index.html` and chrome/sidebar UI but would be a prefix for all modules.

Reply:
I think we want to minimize how much configuration is required. If possible we make it so the web and Misk agree on paths so there’s less rewriting.

Reply:
Agreed 👍